flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/BoundedBlockingSubpartitionTest.java

```diff
@@ -117,6 +121,34 @@ public void testCloseBoundedData() throws Exception {
 		assertTrue(reader.closed);
 	}
 
+	@Test
+	public void testRecycleCurrentBufferOnFailure() throws Exception {
+		final ResultPartition resultPartition = createPartition(ResultPartitionType.BLOCKING, fileChannelManager);
+		final BoundedBlockingSubpartition subpartition = new BoundedBlockingSubpartition(
+				0,
+				resultPartition,
+				new FailingBoundedData(),
+				!sslEnabled && type == BoundedBlockingSubpartitionType.FILE);
+		final BufferConsumer consumer = BufferBuilderTestUtils.createFilledFinishedBufferConsumer(100);
+
+		try {
+			try {
+				subpartition.add(consumer);
+				subpartition.createReadView(new NoOpBufferAvailablityListener());
+				fail("should fail with an exception");
+			} catch (Exception ignored) {
+				// expected
+			}
+
+			assertNotNull(subpartition.getCurrentBuffer());
+			assertFalse(subpartition.getCurrentBuffer().isRecycled());
+		} finally {
+			subpartition.release();
+
+			assertNull(subpartition.getCurrentBuffer());
```

Comment:
The code is not testing if the buffer is really recycled.
`assertTrue(subpartition.getCurrentBuffer().isRecycled())`